File: src/Framework/Output/FFmpegOutput.fs
namespace Animations.Output

open Animations
open System
open System.Diagnostics

type public FFmpegOutput(destinationFilePath: string, properties: AnimationProperties) =
    
    let ffmpegProcess = new Process()

    do
        ffmpegProcess.StartInfo.FileName <- "ffmpeg"
        let argumentsFormat = "-y -f rawvideo -pix_fmt bgra -s:v {0}x{1} -r {2:F2} -i pipe:0 -dst_range 0 -pix_fmt yuv422p -crf 0 \"{3}\""
        ffmpegProcess.StartInfo.Arguments <- String.Format(argumentsFormat, properties.Width, properties.Height, 
                                                           properties.Fps, destinationFilePath)
        ffmpegProcess.StartInfo.UseShellExecute <- false
        ffmpegProcess.StartInfo.RedirectStandardInput <- true
        ffmpegProcess.Start() |> ignore

    interface IDisposable with
        member this.Dispose() = 
            ffmpegProcess.StandardInput.BaseStream.Close()
            ffmpegProcess.WaitForExit()

    member this.Write bytes =
        ffmpegProcess.StandardInput.BaseStream.Write(bytes, 0, bytes.Length)

    interface IVideoOutput with
        member this.Write frame = this.Write frame.Bytes
    

File: Animations.Framework/Output/FFmpegOutput.cs
using System.Diagnostics;
using Animations.Framework.Drawing;

namespace Animations.Framework.Output
{
    public sealed class FFMpegOutput : IVideoOutput
    {
        public FFMpegOutput(string destinationFilePath, AnimationProperties properties)
        {
            _process = new Process();
            _process.StartInfo.FileName = "ffmpeg";
            const string argumentsFormat =
                "-y -f rawvideo -pix_fmt bgra -s:v {0}x{1} -r {2:F2} -i pipe:0 -dst_range 0 -pix_fmt yuv422p -crf 0 \"{3}\"";
            _process.StartInfo.Arguments = string.Format(argumentsFormat, properties.Width, properties.Height,
                properties.Fps, destinationFilePath);
            _process.StartInfo.UseShellExecute = false;
            _process.StartInfo.RedirectStandardInput = true;
            _process.Start();
        }
        
        public void Dispose()
        {
            _process.StandardInput.Close();
            _process.WaitForExit();
        }

        public void Write(Frame frame)
        {
            _process.StandardInput.BaseStream.Write(frame.Bytes, 0, frame.Bytes.Length);
        }

        private readonly Process _process;
    }
}